### DISCUSSION TITLE
Photos Closest City

### DISCUSSION THREAD

**Discussion by oPromessa** [RhetTbull/osxphotos#1666] [Q&A] [answered]
Do you know how Photos 
1. determines the "Closest City" one sees under the Menu 'Image' > option 'Adjust Date and Time...' 
1. and where it saves it in the database
1. and if it is possible to query it via `osxphotos` CLI or API?

__Background__
* While loading some old pics for which I've adjusted (before Import) the date/time with timezone and GPS info via:
   * `exiftool -SubSecDateTimeOriginal="2004:12:03 20:00:00+01:00" file.JPG`
   * `exiftool -GPSLatitude*=10 -GPSLongitude*=10 file.JPG`
* ...I've ended up specifying an incorrect UTCOffset (+01:00) when no DST was valid. It should have been "+00:00".
* Which resulted in Photos finding a closest city in the wrong hemisphere to where the pic was taken.
* So I know Photos makes a combination of GPS position, date/time and TZ info to find "Closest City".Either it does it "on-the-fly" in runtime or it saves it on the database somewhere!

__Why?__
* So I'm now trying to find such pics wrongly setup in the Photos database so that I can fix them and reload them.

__Additional info__
* Already built a [small tool](https://gist.github.com/oPromessa/85c2040a0187fe6868ead8a4b0ed912e) with python libraries pytz and timezonefinder to find the appropriate DST and UTC Offset to properly setup date/time based on GPS Location and date/time via `exiftool` prior to import.
   * Still, I've noticed Photos mostly hits the mark on finding the "Closest City" but sometimes is still a bit off... oh well! I can live with that.

**Comment by RhetTbull** [comment, marked as answer]
The actual closest city as displayed in the Photos GUI is not stored in the database. Presumably this is calling some Apple API in real-time with the coordinates to get the name of the "closest city" (which is really the "closest city in the timezone database"). What is stored in the Photos database is the timezone name from the [IANA timezone database](https://en.wikipedia.org/wiki/Internet_Assigned_Numbers_Authority). Unfortunately osxphotos does not make this available. I'm actually working on some updates to the time-handling code in OSXPhotos at the moment that will expose this but that won't be ready for a while. In the meantime, you can use the following template function to retrieve the information. For example, if you save the following code as "tzname.py", it can be used as a template by running:

`osxphotos query --field uuid "{uuid}" --field filename "{photo.original_filename}" --field timezone "{function:tzname.py::tzname}"`

which will produce a CSV file of UUID, filename, timezone name. 

This isn't exactly what you are looking for but may still allow you to accomplish your goal. For example, if closest city is `Toronto, Canada`, the value retrieved is `America/Toronto`. There may be some instances where only the GMT offset is stored, e.g. `GMT-0400`. 

`tzname.py`

```python
""" Example showing how to use a custom function for osxphotos {function} template
    Use:  osxphotos export /path/to/export --filename "{function:/path/to/tzname.py::tzname}"

    You may place more than one template function in a single file as each is called by name using the {function:file.py::function_name} format
"""

import pathlib
from typing import List, Optional, Union

from osxphotos import PhotoInfo
from osxphotos.phototemplate import RenderOptions


def tzname(
    photo: PhotoInfo, options: RenderOptions, args: Optional[str] = None, **kwargs
) -> Union[List, str]:
    """example function for {function} template; adds suffix of # if photo has adjustments and ! if photo is a favorite

    Args:
        photo: osxphotos.PhotoInfo object
        options: osxphotos.phototemplate.RenderOptions object
        args: optional str of arguments passed to template function
        **kwargs: not currently used, placeholder to keep functions compatible with possible changes to {function}

    Returns:
        timezone name for a photo as a string or empty string if no timezone name
    """

    tzname = photo.tables().ZADDITIONALASSETATTRIBUTES.ZTIMEZONENAME
    return tzname[0] if tzname else ""
```


You may also want to take a look at `osxphotos timewarp` (see `osxphotos help timewarp`) which can bulk adjust the timezones of selected photos.
